---
title: "Lecture1: Introduction to Statistical Genetics with R"
output: rmarkdown::github_document
---

-   2025-03-11

### Computing on a data matrix

* In bioinformatics, gene expression values (from several groups of patients) are stored as rows such that each row contains the expression values of all the patients for a particular gene and each column contains all the gene expression values for a particular person

> 열이 각 gene이고 행이 각 sample, value는 gene expression.

```{r}
gene1 = c(1.00, 1.50, 1.25)
gene1
```

```{r}
gene2 = c(1.35, 1.55, 1.00)
gene3 = c(-1.10, -1.50, -1.25)
gene4 = c(-1.20, -1.30, -1.00)
genes = c(gene1, gene2, gene3, gene4)
genes
```

```{r}
geneData = matrix(genes, nrow=4, ncol=3, byrow=TRUE) # !byrow! -> 벡터를 행으로 박기
geneData
```

```{r}
rownames(geneData) = c("gene1", "gene2", "gene3", "gene4")
colnames(geneData) = c("Eric", "Peter", "Anna")
geneData
# t(geneData)
```

> 여기에선 각 col이 sample이 되고 각 row는 gene name이 된다.

```{r}
# data를 받고 제일 처음 해야할 것.
nrow(geneData) # # of gene
ncol(geneData) # # of sample
class(geneData)
```

```{r}
geneData[3,] # third row (gene) data
geneData[c(1,2), ] # first and second row(gene) data
geneData[c("gene1", "gene2"), ] # can indexing with names
```

```{r}
apply(geneData, 1, mean) # 1 means mean of each row(gene) == row by row
apply(geneData, 2, mean) # 2 means mean of each column(sample) == col by col
```

### Constructing a data frame

-   Matrix must have same data type init.
-   But Data frame do not need it
-   하나의 컬럼에서는 똑같은 tpye이 들어가야함.
-   각 행의 열 수가 같아야함.
-   It consists of a list of column vectors of equal length, where the data in each column must be of the same type.

```{r}
patients.df = data.frame(
  patientID = c("101", "102", "103", "104"),
  treatment = c("drug", "placebo", "drug", "placebo"),
  age = c(20, 30, 24, 22) 
)

patients.df
```



```{r}
nrow(patients.df)
ncol(patients.df)
class(patients.df)
```


### Referencing column vectors in a data frame

```{r}
patients.df[, 2]

patients.df[[2]]
patients.df[["treatment"]]
patients.df$treatment # 이걸 제일 많이 사용함
patients.df[, 2]
```

```{r}
patients.df[c(1,3), ]

patients.df$treatment=="drug"
patients.df[patients.df$treatment=="drug", ] # [, (모든 열)]

subset(patients.df, treatment=="drug")
```

```{r}
row.names(patients.df) = patients.df$patientID
patients.df[c("101", "103"),]
```

```{r}
patients.df[2] # series
patients.df[[2]] # values

patients.df["treatment"]
patients.df[["treatment"]]
```

### Constructing a list

-   list는 매우 자유로움!

```{r}
list1 <- list(
  c("p53", "p63", "p73"),
  matrix(rnorm(10), nrow=2),
  c(TRUE, FALSE, TRUE, FALSE, FALSE)
)
list1
```

```{r}
p53FamilyGenes <- c("p53", "p63", "p73")
list2 <- list(
  genes = p53FamilyGenes,
  geneExpression = matrix(rnorm(10), nrow=2),
  remission = c(TRUE, FALSE, TRUE, FALSE, FALSE)
)
list2
```

```{r}
list2[1]
list2[[1]]
list2[["genes"]]
list2$genes
```

```{r}
list2[[1]][1] <- "her2"
list2$genes
list2
```

```{r}
list2[[2]]
list2$geneExpression
```

```{r}
list2[[3]]
list2$remission
```

### Data path attachment

```{r}
# remission # error
attach(list2) # 데이터 이름 생략하기 위해서..!
remission

detach(list2)
# remission # error
```

```{r}
genes <- c("p53", "p63", "p73")
attach(list2)
genes # 우선순위는 기존 객체..
list2$genes
```

```{r}
patients.df$treatment
attach(patients.df)

treatment
detach(patients.df)
```

### Generating factor

-   For plotting and testing of hypo we need to generate a **factor**.

```{r}
?gl
gl(3, 5) # gl( # of group, # of element of each group) # factor
rep(1:3, each = 5) # vector
```

-   factor type은 범주형 자료라 연산 안됨. (확인해보기)

```{r}
class(gl(3, 5))
class(rep(1:3, each = 5))
```

```{r}
factor(rep(1:3, each=5)) # same as gl()

A <- factor(rep(1:3, each = 5))
class(A)
```

### Functions

```{r}
set.seed(123)

A <- matrix(sample(1:100, 12), nrow=3, ncol=4)

A

apply(A, 2, var)
```

```{r}
fun <- function(x) {
  sum((x-mean(x))^2)/(length(x)-1)
}

apply(A, 2, fun)

apply(A, 2, function(x) sum((x-mean(x))^2)/(length(x) - 1))
```

-   case : more than one argment

```{r}
fun1 <- function(x, a) sum(x-a)^2/(length(x)-1)
apply(A, 2, fun1, a=40)
```

```{r}
fun2 <- function(x, a, b) sum(x - a)^2 / b
apply(A, 2, fun2, a = 40, b = 10)
```

```{r}
list(a=2, b=3)

lapply(list(a=2, b=3), sqrt) # 각각의 원소마다 적용하라..

matrix(1:6, 2, 3)
lapply(matrix(1:6, 2, 3), function(x) x^2) # 각 원소가 함수로 들어가서 lapply이므로 output으로 list로 출력된다.
```

```{r}
mylapply <- function(lst, fun) {
  lst <- as.list(lst)
  ans <- vector("list", length(lst))
  names(ans) <- names(lst)
  for(i in 1:length(lst))
    ans[i] <- list(fun(lst[[i]]))
  ans
  }
mylapply(list(a=2, b = 3), sqrt)
```

### Introduction to Bioconductor

```{r}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install()
```

```{r}
#BiocManager::install("ALL")
library(ALL)
```

```{r}
data(ALL)
#?ALL
ALL
str(ALL)
```

```{r}
class(ALL)
```

```{r}
experimentData(ALL)
#sampleNames(ALL)
#featureNames(ALL)
```

```{r}
dim(exprs(ALL))
exprs(ALL)[1:10, 1:5]
```

```{r}
varLabels(ALL)
ALL$sex
ALL$age
ALL$mol.biol
```

### ALL Data Example

```{r}
ALL$BT

table(ALL$BT)
```

```{r}
#' ALL$BT에서 T0_4까지를 BT로 변환을 위한 작업
#' 다음 코드를 실행하면 T/F의 Logical vector가 생성된다.
BT <- ALL$BT %in% c("T", "T1", "T2", "T3", "T4")

#' as.numeric(BT)를 해서 T/F 값을 0, 1로 변환
#' 그 다음 label을 설정해서 0 -> B 1 -> T로 factor 변환
y <- factor(as.numeric(BT), labels=c("B", "T")) # logical로 처리
table(y)
```

```{r}
x <- t(exprs(ALL))
dim(x)

summary(as.numeric(x)) # 이거 체크 vs summary(x)

hist(x, col="orange", nclass=50, main="Expression Data")
```

-   scaling 할 때 행은 sample 열은 variable로 설정하기!!!!!!!

```{r}
x2 <- scale(x)
dim(x2)
summary(as.numeric(x2))
hist(x2, col="orange", nclass=50, main="Scaled Expression Data")
```


### golub Data example

```{r}
#BiocManager::install("multtest")
library(multtest)
data(golub)
class(golub)
```


```{r}
class(golub.cl)
length(golub.cl)
golub.cl
```


```{r}
class(golub.gnames)
dim(golub.gnames)
head(golub.gnames)
```

```{r}
dim(golub)
golub[1:10, 1:5] # row -> samples , col = genes
```

```{r}
golub.gnames[1042, ]
golub[1042, ]
```


```{r}
golubFactor <- factor(golub.cl, levels=0:1,
                        labels = c("ALL","AML")) # 이런 식으로 factor로 치환하는게 좋다..!
golubFactor
table(golubFactor)

golubFactor=="ALL" 
golub[1042, golubFactor == "ALL"]
```


```{r}
# 각 sample(row)에 대한 ALL gene expression value의 mean
meanALL <- apply(golub[ ,golubFactor=="ALL"], 1, mean)
length(meanALL)
summary(meanALL)
```


```{r}
meanAML <- apply(golub[ ,golubFactor=="AML"], 1, mean)
summary(meanAML)
```


```{r}
par(mfrow=c(1,2))
hist(meanALL, col="orange", nclass=50, main="ALL")
hist(meanAML, col="purple", nclass=50, main="AML")
```


```{r}
ALL <- as.numeric(golub[ ,golubFactor=="ALL"])
AML <- as.numeric(golub[ ,golubFactor=="AML"])
```


```{r}
par(mfrow=c(1,2))
hist(ALL, col="orange", nclass=50, main="ALL")
hist(AML, col="purple", nclass=50, main = "AML")
```

### `grep()`

* We are interested in the properties of certain genes.

* For instance, **gene CD33** plays an important role in distinguishing lymphoid from myeloid lineage cells.

* To perform computations on the expressions of this gene **we need to know its row index**.

> grep 함수는 지정한 문자열("CD33")이 포함된 요소를 찾기 위해 사용, 이 코드는 golub.gnames 객체의 두 번째 열에서 "CD33"이라는 문자열을 (대소문자 구분 없이) 검색하고, 그 문자열이 포함된 행의 인덱스를 반환하여 변수 cd33에 저장.

```{r}
#' grep function
#' ignore.case: TRUE로 설정하면 대소문자 구분 없이 검색
cd33 <- grep("CD33", golub.gnames[ ,2], ignore.case=TRUE)
cd33
golub[cd33, ]
golub.gnames[cd33, ]
```


```{r}
ccnd3 <- grep("CCND3", golub.gnames[ ,2], ignore.case=TRUE)
ccnd3
golub.gnames[ccnd3,]
```

> 즉 조건을 만족하지 않는 행의 인덱스를 찾고 싶다면, grep 함수의 invert 옵션을 사용

```{r}
not_cd33 <- grep("CD33", golub.gnames[,2], ignore.case=TRUE, invert=TRUE)
```


> ```
in python
df[df['column'].str.contains("CD33", case=False)] 
```







